lib/cog/commands/filter.ex

```diff
@@ -26,29 +35,55 @@ defmodule Cog.Commands.Filter do
     {:reply, req.reply_to, result, state}
   end
 
-  defp maybe_filter(item, %{"field" => field, "matches" => matches}) do
-    match = with {:ok, field} <- Access.fetch(item, field),
-      regex = compile_regex(matches),
-      field_string = to_string(field),
-      do: String.match?(field_string, regex)
-
-    case match do
-      true ->
-        item
-      _ ->
-        nil
-    end
+  defp maybe_filter(item, %{"path" => path, "matches" => matches}) do
+    build_path(String.split(path, "."), [], [])
+    |> fetch(item, matches)
   end
-  defp maybe_filter(item, %{"field" => field}) do
-    case Access.fetch(item, field) do
+  defp maybe_filter(item, %{"path" => path}) do
+    build_path(String.split(path, "."), [], [])
+    |> fetch(item)
+  end
+  defp maybe_filter(_item, %{"matches" => _matches}),
+    do: %{error: "Please include a `path` to match with"}
+  defp maybe_filter(item, _),
+    do: item
+
+  defp fetch([], item), do: item
+  defp fetch([key|path], item) do
+    result = case Access.fetch(item, key) do
       {:ok, _} ->
-        item
+        fetch_path(true, path, item, key)
       :error ->
         nil
     end
+    fetch(path, result)
   end
-  defp maybe_filter(item, _),
-    do: item
+
+  defp fetch([], item, matches) do
```

Comment:
I'm finding it very difficult to trace through this section of the code; I think it's trying to do too much at once, and ends up obscuring things.

Conceptually, at least, we want to see if anything exists at the specified path, and if it does (and we've been given a regular expression) see if the value matches. The fact that we're treating values as strings for the purpose of matching might warrant a dedicated function in order to highlight this fact, and to make it easier to modify the behavior if / when we decide to provide data-type-specific matching (e.g., specifying numerical ranges or cutoffs, for example, which are challenging to do correctly with regular expressions).